Create initial documentation (#2)
kubeflow/pipelines-components: docs/CONTRIBUTING.md

- [Testing Requirements](#testing-requirements)
- [Review Process](#review-process)
- [Before Submitting](#before-submitting)
- [Getting Help](#getting-help)
- [Naming Conventions](#naming-conventions)
- [References](#references)

## Quick Start

*Get up and running with your first contribution in 7 simple steps.*

1. **Read the [Kubeflow Community Code of Conduct](https://github.com/kubeflow/community/blob/master/CODE_OF_CONDUCT.md)** - All contributors must follow these guidelines
2. **Fork and clone** the repository
3. **Set up environment** - see [ONBOARDING.md](ONBOARDING.md)
4. **Create feature branch** - `git checkout -b component/my-component`
5. **Develop component** following the structure below
6. **Test thoroughly** - see [TESTING.md](TESTING.md)
7. **Submit pull request**

## What We Accept

*Types of contributions we welcome to the repository.*

- **Individual Components**: ML tasks (preprocessing, training, evaluation)
- **Component Collections**: Related component sets
- **Pipeline Templates**: Complete pipeline examples
- **Bug Fixes**: Improvements to existing components

## Component Structure

*Standard directory layout and files required for every component.*

> Each component must follow this structure:

Reviewer: we have already mentioned this in onboarding right? could we instead link to the onboarding doc instead of repeating it here? Or we can just say in general "read the onboarding doc for getting started"

# Contributing to Kubeflow Pipelines Components

Welcome! This guide covers how to contribute components to the Kubeflow Pipelines Components Repository.

## Table of Contents

- [Quick Start](#quick-start)
- [What We Accept](#what-we-accept)
- [Component Structure](#component-structure)

> - [Required Files](#required-files)

Reviewer 2: Link doesn't work, doesn't look like this heading is present

2. **Fork and clone** the repository
3. **Set up environment** - see [ONBOARDING.md](ONBOARDING.md)
4. **Create feature branch** - `git checkout -b component/my-component`
5. **Develop component** following the structure below
6. **Test thoroughly** - see [TESTING.md](TESTING.md)
7. **Submit pull request**

## What We Accept

*Types of contributions we welcome to the repository.*

- **Individual Components**: ML tasks (preprocessing, training, evaluation)
- **Component Collections**: Related component sets
- **Pipeline Templates**: Complete pipeline examples
- **Bug Fixes**: Improvements to existing components

## Component Structure

*Standard directory layout and files required for every component.*

Each component must follow this structure:

```
components/my_component/   # or third_party/components/my_component
├── __init__.py            # (exposes the component entrypoint for imports)
├── component.py
├── metadata.yaml          # Component specification
├── README.md              # Documentation
├── OWNERS                 # Maintainers

> ├── Containerfile          # Container definition

Reviewer 3: Containerfiles are optional.

cd pipeline-components

# Add upstream remote to sync with main repository
git remote add upstream https://github.com/kubeflow/pipeline-components.git

# Verify remotes are set up correctly
git remote -v
```

### 2. Environment Setup

```bash
# Set up development environment (see ONBOARDING.md for details)
uv venv
source .venv/bin/activate  # On Windows: .venv\Scripts\activate
uv pip install -r requirements-dev.txt

# Install pre-commit hooks for code quality
pre-commit install

# Verify setup works
./scripts/lint.sh
pytest
```

### 3. Create Feature Branch

```bash
# Always sync with upstream before creating new branches
git fetch upstream
git checkout main
git merge upstream/main

# Create descriptive feature branch
git checkout -b component/data-preprocessing
# or for bug fixes:
git checkout -b fix/issue-123
```

### 4. Implement Your Component

Create your component following the required structure:

```bash
# Create component directory
mkdir -p components/my_component/{src,tests}

# Create required files
touch components/my_component/metadata.yaml
touch components/my_component/README.md
touch components/my_component/OWNERS

> touch components/my_component/Containerfile

Reviewer 3: ```suggestion
touch components/my_component/Containerfile # required only for custom images
```

touch components/my_component/src/{main.py,requirements.txt}
touch components/my_component/tests/test_main.py
```

**Implementation checklist:**
- [ ] Write component logic in `src/main.py`
- [ ] Define dependencies in `src/requirements.txt`
- [ ] Create comprehensive `metadata.yaml`
- [ ] Write detailed `README.md` with examples
- [ ] Add maintainers to `OWNERS` file

> - [ ] Build efficient `Containerfile`

Reviewer 3: ```suggestion
- [ ] Build efficient `Containerfile` (only if custom image required)
```

- **Integration tests**: Test end-to-end functionality  
- **Coverage**: Minimum 80%
- **Error handling**: Test failure scenarios

See [TESTING.md](TESTING.md) for detailed testing guidelines.

## Review Process

*How contributions are reviewed and what criteria they must meet.*

All PRs must pass:
- Automated checks (linting, tests, builds)
- Code review by maintainers
- Documentation review

### Review Criteria

- Component works as described
- Code is clean and well-documented
- Tests provide good coverage
- Follows repository standards

## Before Submitting

*Final checklist to ensure your contribution is ready for review.*

- [ ] Component builds successfully
- [ ] All tests pass (≥80% coverage)
- [ ] All required files present
- [ ] Documentation complete
- [ ] Follows naming conventions
- [ ] No security vulnerabilities

## Getting Help

*Resources and channels for getting support during development.*

- **Setup**: [ONBOARDING.md](ONBOARDING.md)
- **Testing**: [TESTING.md](TESTING.md)
- **Governance**: [GOVERNANCE.md](GOVERNANCE.md)
- **Best Practices**: [BESTPRACTICES.md](BESTPRACTICES.md) *(coming soon)*
- **Agents**: [AGENTS.md](AGENTS.md) *(coming soon)*

> - **Community**: [#kubeflow-pipelines Slack](https://kubeflow.slack.com/channels/kubeflow-pipelines)

Reviewer 3: This is the old KFP Slack. If we point to the website as a canonical link, we prevent it from being outdated in the future.

```suggestion
- **Community**: `#kubeflow-pipelines` channel on the [CNCF Slack](https://www.kubeflow.org/docs/about/community/#kubeflow-slack-channels)
```

cd pipelines-components
git remote add upstream https://github.com/kubeflow/pipelines-components.git

# Set up Python environment
uv venv
source .venv/bin/activate
uv pip install -r requirements-dev.txt

# Install pre-commit hooks for automatic code quality checks
pre-commit install

# Verify your setup works
pytest
```

## What We Accept

We welcome contributions of production-ready ML components and re-usable pipelines:

- **Components** are individual ML tasks (data processing, training, evaluation, deployment) with usage examples
- **Pipelines** are complete multi-step workflows that can be nested within other pipelines
- **Bug fixes** improve existing components or fix documentation issues

## Component Structure

Components must be organized by category under `components/<category>/` (Core tier) or `third_party/components/<category>/` (Third-Party tier).

Pipelines must be organized by category under `pipelines/<category>/` (Core tier) or `third_party/pipelines/<category>/` (Third-Party tier).

## Naming Conventions

- **Components and pipelines** use `snake_case` (e.g., `data_preprocessing`, `model_trainer`)

> - **Git branches** use `component/name` for features or `fix/issue-123` for bug fixes

Reviewer: is this for user branches during PRs? we don't need to enforce a rule here

├── example_pipelines.py   # Working usage examples
└── tests/
│   └── test_pipeline.py  # Unit tests
└── <supporting_files>
```

### metadata.yaml Schema

Your `metadata.yaml` must include these fields:

```yaml
tier: core  # or 'third_party'
name: my_component
stability: stable  # 'alpha', 'beta', or 'stable'
dependencies:
  kubeflow:
    - name: Pipelines
      version: '>=2.5'
  external_services:  # Optional list of external dependencies
    - name: Argo Workflows
      version: "3.6"
tags:  # Optional keywords for discoverability
  - training
  - evaluation
lastVerified: 2025-11-18T00:00:00Z  # Updated annually; components are removed after 12 months without update
ci:
  compile_check: true  # Validates component compiles with kfp.compiler
  skip_dependency_probe: false  # Set true only with justification
  pytest: optional  # Set to 'required' for Core tier
packages_to_install:  # Dependencies verified by CI
  - pandas>=1.5.0
  - scikit-learn>=1.2.0
links:
  documentation: https://kubeflow.org/components/my_component
  issue_tracker: https://github.com/kubeflow/pipelines-components/issues
```

### OWNERS File

> The OWNERS file enables component owners to self-service maintenance tasks including approvals, metadata updates, and lifecycle management:

Reviewer: can you either link to the [k8s docs](https://docs.prow.k8s.io/docs/components/plugins/approve/approvers/#lgtm-label) or add a brief description on `/lgtm `and `/approve` semantics

as I believe we intend to use prow like the rest of the kf projects

git commit -m "feat(training): add <my_component> training component

- Implements <my_component> Core-Tier component
- Includes comprehensive unit tests with 95% coverage
- Provides working pipeline examples
- Resolves #123"
```

### Push and Create Pull Request

Push your changes and create a pull request on GitHub:

```bash
git push origin component/my-component
```

On GitHub, click "Compare & pull request" and fill out the PR template provided with appropriate details

All PRs must pass:
- Automated checks (linting, tests, builds)
- Code review by maintainers and community members
- Documentation review

### Pre-Submission Checklist

Verify your contribution meets all requirements before requesting review:

- [ ] All tests pass
- [ ] Pre-commit hooks pass without errors
- [ ] All required files are present and complete
- [ ] `metadata.yaml` includes fresh `lastVerified` timestamp
- [ ] OWNERS file lists appropriate maintainers
- [ ] README provides clear documentation with usage examples
- [ ] Component follows `snake_case` naming convention
- [ ] (Core tier only) Containerfile present if using custom image
- [ ] No security vulnerabilities in dependencies

> - [ ] Container builds successfully (if applicable)

Reviewer: this should be a checklist in the PR template it self and not added here, otherwise we have to update two places everytime the pr template is updated

Reviewer: okay, just ensure it's captured in some Jira or somewhere so we don't forget
